### PR TITLE
Integrate basic braille viewer

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
#7788

### Summary of the issue:
Having a braille viewer tool integrated into NVDA is quite helpful for developers who wish to test with braille but do not have a braille device. This tool may also be useful when two people are using a computer, one with braille and one sighted (for instance a parent, reading with their child)

### Description of how this pull request fixes the issue:
Introduces a simple braille viewer tool.

### Testing performed:
Tested permutations of:
 - Enable on startup true / false
- Starting / exiting NVDA
- Toggling the braille viewer from the tools menu.

As per https://github.com/nvaccess/nvda/issues/7788#issuecomment-399048734 tested an empty message no longer results in an error.

**Not tested**
Using the brailleViewer in conjunction with another display. I expect this to work, and the output should match.

### Known issues with pull request:
Scrolling gestures (braille_scroll_forward, braille_scroll_back) must be set manually, there are no buttons in the GUI to do this. Scrolling with buttons on the GUI will be quite a lot of work and won't be covered in this PR.

Routing is not supported, for similar reasons to buttons on the GUI for scrolling.

- [x] User guide needs to be updated

### Change log entry:
New features:
```
- Braille Viewer tool, allows viewing braille output via an on-screen window. (#7788)
```

### CODE REVIEW

**Comment by leonardder** [comment]
Could you have a look at the extension points I created in the braille module as part of #9917? In short:

* braille.handler.pre_writeCells could be used to update the braille viewer's contents
* braille.handler.filter_displaySize can be used to set the display size when there is no display connected. Optionally, the braille viewer can control the number of cells that are send to a physical display if the physical display is too big for the contents of the braille viewer to fit on screen.

note that in its current form, the braille viewer is very likely to how braille is handled in NVDA Remote. Both would be able to work seamlessly with these extension points.

**Comment by feerrenrut** [comment]
> note that in its current form, the braille viewer is very likely to conflict with how braille is handled in NVDA Remote. 

To make sure I understand, is this the case for the remote server, or the client? For these extension points to fix the problem you are describing, NVDA remote will need to be updated, correct?

Given that NVDA remote currently doesn't work on alpha, I think I would prefer to merge this first, and re-address it when we look at compatibility with NVDA Remote. This may involve making braille viewer use these extension points, but as it stands using extension points for this is overkill IMO.

**Comment by leonardder** [comment]
> To make sure I understand, is this the case for the remote server, or the client? 

I assume you're talking about the controlled vs. controlling machine here? In that case, both, as both send updates to the other machines when the display (not the display size) is changed. The smallest display size will be applied to all systems that share a connection.

> For these extension points to fix the problem you are describing, NVDA remote will need to be updated, correct?

Yes.

> Given that NVDA remote currently doesn't work on alpha, I think I would prefer to merge this first, and re-address it when we look at compatibility with NVDA Remote. This may involve making braille viewer use these extension points, but as it stands using extension points for this is overkill IMO.

This is somehow correct. Another conflict would be introduced with the ALPHA driver, which has split point functionality that changes the braille.handler.displaySize. However, that does not update the size of the braille viewer.

I"m also worried that in the current situation, the following can happen:

1. You have a 40 cell display connected
2. Enable the braille viewer
3. Disconnect the display
4. The braille viewer is stuck, since disconnecting the braille display disables the braille handler.

I think this can be solved by reevaluating the state of the braille viewer in braille.handler.setDisplayByName. However, this is also covered by the extension points, as the enabled state of the braille handler will be dependent on display size and no longer an independent attribute.

**Comment by JulienCochuyt** [comment]
Thank you very much for this long awaited feature!

> **Not tested**
> Using the brailleViewer in conjunction with another display. I expect this to work, and the output should match.

Tested from source with both 40 and 80 cells autodetected devices.

Test 1: No braille device attached.
Results:
- Braille viewer shows 40 cells as expected.
- Two lines are displayed, one with braille characters and one with printed characters. Both lines are not aligned. The use of a fixed-width font for the second line could solve this.
- The checkbox for enabling the braille viewer on startup makes the window use more height as one could need for a wide always on top window. Maybe this could go in the settings dialog along with an option to display either line 1, line 2, or both.
- The caret position appears on the braille line only, not on the printed text line. This could probably be improved by using - if not caret glyphs - different colors for the review and system carets.

Test 2: Starting with the 40 cells device attached.
Result: OK

Test 3: Starting with the 80 cells device attached.
Result: OK - The braille viewer window doubles in width to display all 80 characters.

Test 4: Starting with a device attached, then unplugging it.
Result: As @leonardder predicted, the braille viewer stops functioning, stuck on what was displayed before unplugging the device.

Test 5: Starting with the 80 cells display attached, then unplugging it and attaching the 40 cells display.
Result: Works, except that the double sized braille viewer window does not shrink.

Test 6: Starting with the 40 cells display attached, then unplugging it and attaching the 80 cells display.
Result: Somwhat works, but half of the display is not shown as the window does not resize to double width.


Regarding the use of extension points and compatibility with a future revamping of the Remote Access feature (be it as part of NVDA core or as an add-on): Despite the Braille Viewer being a long awaited feature - much expected by viewing accessibility testers - I would advocate that the lack of a Remote Access feature in the next major release is a show stopper for many users. As such, I think that IMHO the implementation merged into master should be as compatible as we can anticipate will be needed for the Remote Access feature to come.

**Comment by dingpengyu** [comment]
hi @feerrenrut 
Can you fix the build error for All checks have failed, I can't get the latest test build?
thank

**Comment by feerrenrut** [comment]
@JulienCochuyt thanks for testing this.
> Both lines are not aligned. The use of a fixed-width font for the second line could solve this.

I'm aware of this issue, though it has proved difficult to fix. I am setting the font to "Courier New", and the family to `wx.FONTFAMILY_TELETYPE` ([see `setFont`](https://github.com/nvaccess/nvda/blob/6426ade03c80506c8ee3e7886101caabf8f8b550/source/brailleViewer/brailleViewerGui.py#L104)), but the raw text characters do not align with the braille. I have set the size of the two fonts so they are closer. I think that to display the braille characters there is a fallback to some other font perhaps? Even if we make this work well with latin based characters, it will likely be broken in other character sets. There is a also the complication that sometimes certain single character raw text translates to several braille characters. These issues might be solvable, but I think they should be looked at in the future. 

> The checkbox for enabling the braille viewer on startup makes the window use more height as one could need for a wide always on top window.

For now I will leave it as it is. It matches the speech viewer, if we move this option for the braille viewer, I think we should do the same for speech viewer. 

> As @leonardder predicted, the braille viewer stops functioning, stuck on what was displayed before unplugging the device.

Thanks, I will fix this up shortly. 

> the implementation merged into master should be as compatible as we can anticipate 

I don't think the current implementation will cause problems for making NVDA remote compatible. We may have to change this implementation at that stage, but to do so now would just be guess work.

File: source/braille.py
Comment (leonardder) [inline source/braille.py]: Should this be logged?

File: source/braille.py
Comment (leonardder) [inline source/braille.py]: Note that Braille is initialized before and terminated after the gui. When gui is terminated before braille, does this call actually make sense?

I'm also worried that resetting the configuration to currently saved config will terminate the braille viewer and not initialize it again. As the braille viewer does not talk to the braille handler apart from at initialization time, I think it is safe to leave this out.

File: source/braille.py
Comment (leonardder) [inline source/braille.py]: I don't think displaySize should be an optional:

```suggestion
			self.displaySize = 0 if not self.display else self.display.numCells
```

File: source/brailleDisplayDrivers/noBraille.py
Comment (leonardder) [inline source/brailleDisplayDrivers/noBraille.py]: Interesting approach. I think I'd rather see setDisplayByName update the braille viewers state, including the size of the window. I don't think the current approach solves everything @JulienCochuyt came up with.

File: source/brailleViewer/__init__.py
Comment (leonardder) [inline source/brailleViewer/__init__.py]: May be the comment should also tell how it is destroyed in this case. Now it only tells how it is not.

File: source/brailleViewer/__init__.py
Comment (leonardder) [inline source/brailleViewer/__init__.py]: I'm worried that this might happen if the gui is initialized after braille.

File: source/braille.py
Comment (leonardder) [inline source/braille.py]: I"m worried that this could somehow slow down braille output, as braille output is blocked by the braille viewer being updated. Have you considered rate limiting here?

File: source/config/configSpec.py
Comment (leonardder) [inline source/config/configSpec.py]: This flag is no longer used

```suggestion
```

File: source/gui/__init__.py
Comment (leonardder) [inline source/gui/__init__.py]: I think this has to be moved two lines below. This is now causing checkpot to fail.

File: user_docs/en/userGuide.t2t
Comment (leonardder) [inline user_docs/en/userGuide.t2t]: I think this line might be better to be more at the start of the braille viewer section.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

File: source/braille.py
Comment (feerrenrut) [reply source/braille.py]: I'd like to see that there is a problem to solve first. Bear in mind, most people wont be using the braille viewer and a display. If testing shows there is a problem, then we have a benchmark that we can try to improve on.

File: source/brailleViewer/__init__.py
Comment (feerrenrut) [reply source/brailleViewer/__init__.py]: braille doesn't call createBrailleViewer, the GUI does. I don't think there is anything to worry about right now.

File: source/braille.py
Comment (leonardder) [inline source/braille.py]: Note that this is at least set by the alva driver in core when changing the split point, and then it will break this. You will have to change the alva driver to use braille.handler._displaySize instead.

File: source/brailleViewer/__init__.py
Comment (leonardder) [inline source/brailleViewer/__init__.py]: It worries me slightly that this sets numCells according to braille.handler.displaySize, which itself also checks the braille viewer. I think it will work as expected, though.

File: source/brailleViewer/brailleViewerGui.py
Comment (leonardder) [inline source/brailleViewer/brailleViewerGui.py]: Please add appropriate copyright headers.

File: source/brailleViewer/brailleViewerGui.py
Comment (leonardder) [inline source/brailleViewer/brailleViewerGui.py]: May be make this comment less personal 😂

File: source/config/configSpec.py
Comment (leonardder) [inline source/config/configSpec.py]: I don't think this is used. Must have missed this before.

File: source/brailleViewer/brailleViewerGui.py
Comment (leonardder) [inline source/brailleViewer/brailleViewerGui.py]: It is a bit confusing that this is a parameter to the class constructor. It is only used to create an initial row of empty braille cells. Is it really necessary to keep it?
If so, I'd rather have the main frame calculate the number of braille cells to use.

Comment (feerrenrut) [reply source/brailleViewer/brailleViewerGui.py]: It allows the GUI to initialise to the right size. I think I would prefer to keep it.

File: source/brailleViewer/__init__.py
Comment (feerrenrut) [reply source/brailleViewer/__init__.py]: Yes, this can now be simplified.

File: source/brailleViewer/brailleViewerGui.py
Comment (leonardder) [reply source/brailleViewer/brailleViewerGui.py]: Doesn't the gui require a reinit if you switch from no display with braille viewer on to a display with 80 cells?

**Review by LeonarddeR — APPROVED** [review]
Apart from my comment about resizing the gui when going from no display with braille viewer (40 cells) to a display with 80 cells, this looks ok to me.

**Comment by DrSooom** [comment]
Why do you don't use Consolas for both lines? Consolas is much easier to read instead of Courier New.

**Comment by feerrenrut** [comment]
> Why do you don't use Consolas for both lines? Consolas is much easier to read instead of Courier New.

Mostly this came down to trying to find a font that would align the braille characters with the raw text characters. At the time Courier New was the closest match I could find. I have done some further investigation on this, I think that the braille was always falling back to using Segoe UI font. Instead, I have included a custom font, which has a monospace variation with Braille symbol support. This font, "Free mono", had some it's own issues with braille symbols, but being GPL3 I was able to modify and fix them. Because we are using the same monospace font for both the braille and the raw text, the widths of the characters will match, reducing the number of things that will make them out of alignment. There is still the problem of cells used to indicate numbers or capitals (excuse my ignorance of terminology here please), but I think these should be left as a problem to solve another day.

@leonardder @JulienCochuyt I think I have fixed the issues with changing the number of cells now. If you are able to test this again please? I would like to merge this next week.

File: source/gui/__init__.py
Comment (leonardder) [inline source/gui/__init__.py]: I don't think this line is necessary.

**Review by LeonarddeR — APPROVED** [review]
I think overall, this looks ok.
Re alligning the braille dots with the text, this shouldn't be that difficult when using brailleToRawPos and rawToBraillePos properly. Makes sense to do that in a follow up pr, though

**Pull request merged**
